lib/compress/zstd_compress.c

```diff
@@ -2566,6 +2567,26 @@ size_t ZSTD_getSequences(ZSTD_CCtx* zc, ZSTD_Sequence* outSeqs,
 
     ZSTD_compress2(zc, dst, dstCapacity, src, srcSize);
     ZSTD_customFree(dst, ZSTD_defaultCMem);
+
+    if (format == ZSTD_sf_noBlockDelimiters) {
+        size_t i = 0;
+        size_t totalSeqs = zc->seqCollector.seqIndex;
+        while(i < totalSeqs) {
+            if (seqCollector.seqStart[i].offset == 0 && seqCollector.seqStart[i].matchLength == 0) {
+                /* Merge the block boundary or last literals */
+                if (i != totalSeqs-1) {
+                    /* Add last literals to next sequence, then "delete" this sequence */
+                    seqCollector.seqStart[i+1].litLength += seqCollector.seqStart[i].litLength;
+                    ZSTD_memmove(seqCollector.seqStart+i, seqCollector.seqStart+i+1, (totalSeqs-i-1)*sizeof(ZSTD_Sequence));
```

Comment:
This is quadratic in the worst case.

A better way to do this is to have have an input index and an output index. Something like:
```c
while (input < totalSeqs) {
  if (merge) {
    seqCollector.seqStart[output] = seqCollector.seqStart[input+1];
    seqCollector.seqStart[output].litLength += seqCollector.seqStart[input];
    output += 1;
    input += 2;
  } else {
    seqCollector.seqStart[output++] = seqCollector.seqStart[input++];
  }
}
```

Reply:
Yeah lol, the naive approach took way too much time even just on `silesia.tar`.